### PR TITLE
Fix Incorrect Hidden State Extraction with Right Padding

### PR DESCRIPTION
**Pull request by xiuyuz**
# Fix Incorrect Hidden State Extraction with Right Padding

## Summary
This PR fixes a bug in `models.py` where the model was incorrectly extracting hidden states from padding tokens when using right-padded batches.

## Bug Description
In `generate_latent_batch` and `generate_latent_batch_hidden_state`, the code used `[:, -1, :]` to extract the hidden states of the last token in the sequence:
```python
e_t = outputs.hidden_states[0][:, -1, :]
last_hidden = outputs.hidden_states[-1][:, -1, :]
```
When using batch generation with right padding (e.g., `[token_A, token_B, PAD]`), index `-1` corresponds to the `PAD` token. As a result, the latent reasoning steps were being initialized with the hidden state of the padding token rather than the last actual token of the prompt.

## Fix
The fix uses the `attention_mask` to determine the index of the last non-padding token for each sequence in the batch:
```python
last_token_indices = attention_mask.sum(1) - 1
last_hidden = outputs.hidden_states[-1][torch.arange(batch_size), last_token_indices, :]
```
This ensures that `last_hidden` (and `e_t`) correctly corresponds to the last real token (e.g., `token_B`), regardless of the padding.

### CODE REVIEW

**Comment by lukasmittag** [comment]
any updates here? Had issues with that as well :) so would appreciate the merged fix. For now I just copy pasted.
